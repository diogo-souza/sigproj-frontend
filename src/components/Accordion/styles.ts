import styled, { css, DefaultTheme } from 'styled-components';

type AccordionProps = {
  isOpen: boolean;
};

export const Header = styled.button`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    text-align: start;
    position: relative;
    justify-content: space-between;
    padding: ${theme.spacings.small};

    color: ${theme.colors.titleBlack};
    font-size: ${theme.font.sizes.large};

    border: none;
    background-color: inherit;
    > svg {
      transition: transform 0.6s;
    }
  `}
`;

export const Content = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.titleBlack};
    font-size: ${theme.font.sizes.large};
    font-weight: ${theme.font.semiBold};
    overflow: hidden;
    max-height: 0;
    opacity: 0;
    pointer-events: none;
    transition: padding 0.6s ease, opacity 0.6s ease, margin 0.6s ease;
  `}
`;

const AccordionModifiers = {
  isOpen: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.white};
    border: 1px solid rgba(18, 18, 18, 0.3);

    ${Header} {
      border-bottom: 1px solid rgba(120, 120, 120, 0.3);
      > svg {
        transform: rotate(90deg);
      }
    }
    ${Content} {
      opacity: 1;
      max-height: 100%;
      pointer-events: all;
      padding: ${theme.spacings.xsmall} ${theme.spacings.small};
      p,
      ul {
        margin: ${theme.spacings.xsmall} 0;
      }

      ul {
        padding-left: ${theme.spacings.small};
      }
    }
  `,
};

export const Container = styled.main<AccordionProps>`
  ${({ theme, isOpen }) => css`
    display: flex;
    flex-direction: column;

    & + & {
      border-top: 1px solid rgba(18, 18, 18, 0.3);
    }

    transition: background-color 0.6s;
    &:hover {
      border: 1px solid rgba(18, 18, 18, 0.3);
      background-color: ${theme.colors.white};
    }

    ${isOpen && AccordionModifiers.isOpen(theme)}
  `}
`;
