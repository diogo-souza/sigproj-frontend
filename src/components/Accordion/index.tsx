import React, { useState } from 'react';
import { FaChevronRight as ArrowRightIcon } from 'react-icons/fa';
import * as S from './styles';

type AccordionProps = {
  question: string;
  answer: string;
};

const Accordion: React.FC<AccordionProps> = ({
  question,
  answer,
}: AccordionProps) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleToggleAccordion = () => {
    setIsOpen(!isOpen);
  };

  return (
    <S.Container isOpen={isOpen}>
      <S.Header
        data-testid="accordion-header"
        onClick={() => handleToggleAccordion()}
      >
        <h3>{question}</h3>
        <ArrowRightIcon />
      </S.Header>

      <S.Content
        data-testid="accordion-content"
        aria-hidden={!isOpen}
        dangerouslySetInnerHTML={{ __html: answer }}
      />
    </S.Container>
  );
};

export default Accordion;
