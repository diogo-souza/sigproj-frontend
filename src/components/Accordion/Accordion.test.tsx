import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';

import Accordion from '.';

describe('<Accordion />', () => {
  it('should render the Accordion', () => {
    renderWithTheme(
      <Accordion question="Pergunta teste" answer="Resposta teste" />,
    );
    expect(screen.getByRole('main')).toBeInTheDocument();
  });
  it('should handle the open/close Accordion', () => {
    renderWithTheme(
      <Accordion question="Pergunta teste" answer="Resposta teste" />,
    );

    // Seleciona o Content
    const fullAccordionElement = screen.getByTestId('accordion-content');

    // Verifica se o Content está escondido
    expect(fullAccordionElement.getAttribute('aria-hidden')).toBe('true');
    expect(fullAccordionElement).toHaveStyle({
      opacity: 0,
      'pointer-events': 'none',
    });

    // Clica no botão de abri menu e verifica se ele abriu
    fireEvent.click(screen.getByTestId('accordion-header'));
    expect(fullAccordionElement.getAttribute('aria-hidden')).toBe('false');
    expect(fullAccordionElement).toHaveStyle({
      opacity: 1,
      'pointer-events': 'all',
    });

    // Clica no botão de fechar Content e verifica se ele fechou
    fireEvent.click(screen.getByTestId('accordion-header'));
    expect(fullAccordionElement.getAttribute('aria-hidden')).toBe('true');
    expect(fullAccordionElement).toHaveStyle({
      opacity: 0,
      'pointer-events': 'none',
    });
  });
});
