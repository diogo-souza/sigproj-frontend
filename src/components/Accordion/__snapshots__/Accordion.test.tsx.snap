// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should render the Accordion 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: start;
  position: relative;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 2.4rem;
  color: #171923;
  font-size: 1.8rem;
  border: none;
  background-color: inherit;
}

.c2 > svg {
  -webkit-transition: -webkit-transform 0.6s;
  -webkit-transition: transform 0.6s;
  transition: transform 0.6s;
}

.c3 {
  color: #171923;
  font-size: 1.8rem;
  font-weight: 500;
  overflow: hidden;
  max-height: 0;
  opacity: 0;
  pointer-events: none;
  -webkit-transition: padding 0.6s ease,opacity 0.6s ease,margin 0.6s ease;
  transition: padding 0.6s ease,opacity 0.6s ease,margin 0.6s ease;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: background-color 0.6s;
  transition: background-color 0.6s;
}

.c0 + .c0 {
  border-top: 1px solid rgba(18,18,18,0.3);
}

.c1:hover {
  border: 1px solid rgba(18,18,18,0.3);
  background-color: #FFFFFF;
}

<main
  class="c0 c1"
>
  <button
    class="c2"
    data-testid="accordion-header"
  >
    <h3>
      Pergunta teste
    </h3>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 320 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
      />
    </svg>
  </button>
  <div
    aria-hidden="true"
    class="c3"
    data-testid="accordion-content"
  >
    Resposta teste
  </div>
</main>
`;
